Add usage of letters for rendering FP SVG.

pszulczewski opened the pull request:
This draft adds only SVG rendering options to use letters or numbers for the grid numbering axis labels.
Selecting letters for the X axis will make the FP look like a chessboard.
![obraz](https://github.com/nautobot/nautobot-app-floor-plan/assets/58683518/a8029a6f-1a03-4ac7-8110-e2ea12d5e3ce)

The downside is that for adding tiles it still uses numbers, because `FloorPlanTile` model `x_origin` and `y_origin` fields are now `PositiveSmallIntegerField`.

![obraz](https://github.com/nautobot/nautobot-app-floor-plan/assets/58683518/f84b1672-c548-4b0e-a538-322d926fa3be)

Potentially form could be changed to present numbers or letters for tile origins, and under clean() it could be covered back to numbers.

joewesch commented:
> Potentially form could be changed to present numbers or letters for tile origins, and under clean() it could be covered back to numbers.

I like this idea. Let's do it!

pszulczewski commented on nautobot_floor_plan/svg.py:
```diff
@@ -81,6 +81,18 @@ def _setup_drawing(self, width, depth):
         return drawing
 
+    @staticmethod
+    def _col_num_to_letter(col_num):
```
Sure, it's still in draft, in other words in conceptual phase.
That may land in helpers.py as it will need more functions to convert letters into numbers and back.

itdependsnetworks commented:
I had mentioned that we should consider having a generator for the increment https://github.com/nautobot/nautobot-app-floor-plan/issues/8#issuecomment-1423580189

pszulczewski commented:
The latest commit is to support letters in the tile form.

![obraz](https://github.com/nautobot/nautobot-app-floor-plan/assets/58683518/5424be16-a6d8-4d0c-9b50-ee17693e6152)

@itdependsnetworks I am fine with implementing generator if @joewesch allows more time.

joewesch commented:
> I had mentioned that we should consider having a generator for the increment

I like the idea, but I think what we have so far will be a good start as I believe it covers at least 80% of use cases. Let's count this as incremental progress toward the eventual goal and not mark this as closing/resolving the issue.

pszulczewski commented:
I am still thinking if keeping x/y grid setting per object will give good user experience? Maybe it should be a global plugin setting, which is selected only once?

joewesch commented on nautobot_floor_plan/forms.py:
```diff
@@ -73,6 +78,7 @@ class FloorPlanTileForm(NautobotModelForm):
     rack = DynamicModelChoiceField(
         queryset=Rack.objects.all(), required=False, query_params={"nautobot_floor_plan_floor_plan": "$floor_plan"}
     )
+    _char = forms.CharField()  # temp char field for replacement in __init__ if grid numbering is using letters.
```
Would it be better to just change the fields to CharField and cast to integers if the choice is `NUMBERS`?

pszulczewski replied:
Currently the form allows only integers. If the field type is not changed, then it comes blank.
Changing to CharField may simplify that. I would need to check

pszulczewski replied:
Yes, that helped.

joewesch commented on nautobot_floor_plan/forms.py:
```diff
@@ -89,3 +95,29 @@ class Meta:
+    def clean(self):
+        """Change x/y origins back to numbers."""
+        if self.x_letters:
+            self.cleaned_data["x_origin"] = utils.column_letter_to_num(self.cleaned_data.get("x_origin"))
```
Should we split these out to proper `clean_x_origin` and `clean_y_origin` methods?

pszulczewski replied:
I started with that, but I was facing some exception because clean() seems to be called before.
Maybe changing the form field to CharField will resolve that.

pszulczewski replied:
Yes, after simplifying form to use just CharFields, clean_ method for each fields just work fine.

joewesch commented:
> I am still thinking if keeping x/y grid setting per object will give good user experience? Maybe it should be a global plugin setting, which is selected only once?

Yes, a plugin setting default would be great, but still allow for grid specific override. Our client uses different grid patterns per Location.

pszulczewski commented on nautobot_floor_plan/forms.py:
```diff
@@ -42,6 +43,12 @@ class Meta:
             "tags",
         ]
 
+    def __init__(self, *args, **kwargs):
+        """Overwrite the constructor to set initial values for select widget."""
+        super().__init__(*args, **kwargs)
+        self.initial["x_axis_labels"] = get_app_settings_or_config("nautobot_floor_plan", "grid_x_axis_labels")
+        self.initial["y_axis_labels"] = get_app_settings_or_config("nautobot_floor_plan", "grid_y_axis_labels")
+
```
`initial=` in the class attributes doesn't seem to work. That's the only solution I found to set the default widget values.

pszulczewski commented:
What is the changelog in the CI? I am not sure why it fails?

joewesch commented:
> What is the changelog in the CI? I am not sure why it fails?

This repo uses towncrier and requires a changelog fragment file: https://docs.nautobot.com/projects/floor-plan/en/latest/dev/contributing/#creating-changelog-fragments

pszulczewski commented:
> > What is the changelog in the CI? I am not sure why it fails?
> 
> This repo uses towncrier and requires a changelog fragment file: https://docs.nautobot.com/projects/floor-plan/en/latest/dev/contributing/#creating-changelog-fragments

Thanks. Looks like my branch was checked out beginning of February and it didn't have `changed` directory.
I've just re-based on `develop` and added a change file.

joewesch commented on nautobot_floor_plan/models.py:
```diff
@@ -53,6 +53,18 @@ class FloorPlan(PrimaryModel):
+        max_length=10,
+        choices=AxisLabelsChoices,
+        default=AxisLabelsChoices.NUMBERS,
+        help_text="Grid numbering of X axis (horizontal).",
```
```suggestion
        help_text="Grid labels of X axis (horizontal).",
```

joewesch commented on nautobot_floor_plan/models.py:
```diff
@@ -53,6 +53,18 @@ class FloorPlan(PrimaryModel):
+        max_length=10,
+        choices=AxisLabelsChoices,
+        default=AxisLabelsChoices.NUMBERS,
+        help_text="Grid numbering of Y axis (vertical).",
```
```suggestion
        help_text="Grid labels of Y axis (vertical).",
```

joewesch commented on docs/admin/install.md:
```diff
@@ -34,6 +34,13 @@ Once installed, the app needs to be enabled in your Nautobot configuration. The
 # In your nautobot_config.py
 PLUGINS = ["nautobot_floor_plan"]
+
+# Optionally you can override default settings for config items to make grid numbering like a chessboard (as seen in this example)
```
```suggestion
# Optionally you can override default settings for config items to make grid labels like a chessboard (as seen in this example)
```

joewesch commented:
Patryk, something isn't working right with the edit form. If I set it to letters and then edit it again, it defaults back to numbers.

joewesch commented:
When setting a label to use letters, but then putting a number in the form it raise an inconsistent error message:
![Screenshot 2024-03-25 at 12 14 50 PM](https://github.com/nautobot/nautobot-app-floor-plan/assets/10467633/975422bc-c806-4e0a-8a2c-d124ce5a3be2)

As part of the clean method we should do one of the following:
- Check for an inconsistent type and raise the appropriate error
- Automatically cast the number or letter to the proper value (my preferred option - save people from their own mistakes)

pszulczewski commented:
Let me check that.

pszulczewski commented:
> Patryk, something isn't working right with the edit form. If I set one (or both) of the labels to letters and then edit the floor plan, the form field defaults back to numbers.

Looks like I stopped working after changing form field to CharField.
I fixed that and added validations. Should be good now, can you please pull and re-test?

joewesch commented on nautobot_floor_plan/forms.py:
```diff
@@ -87,3 +104,48 @@ class Meta:
+            if self.x_letters:
+                self.initial["x_origin"] = utils.col_num_to_letter(self.instance.x_origin)
+            if self.y_letters:
+                self.initial["x_origin"] = utils.col_num_to_letter(self.instance.y_origin)
```
```suggestion
                self.initial["y_origin"] = utils.col_num_to_letter(self.instance.y_origin)
```

pszulczewski replied:
🙈 fixed

pszulczewski commented on nautobot_floor_plan/utils.py:
```diff
@@ -0,0 +1,22 @@
+"""Utilities module."""
+
+
+def col_num_to_letter(number):
```
Done

pszulczewski commented:
@joewesch I have added unittests for the `FloorPlanTileForm` because there weren't any for that form, and they revealed that there were still some issues. I've made the necessary updates in forms, we should be good now.

joewesch commented on docs/admin/install.md:
```diff
@@ -53,3 +60,12 @@ sudo systemctl restart nautobot nautobot-worker nautobot-scheduler
+
+| Key                | Example   | Default  | Description                                                                                                                                    |
+|--------------------|-----------|----------|------------------------------------------------------------------------------------------------------------------------------------------------|
+| default_x_axis_labels | "leters"  | "numbers" | Label style for the floor plan gird. Can use numbers or letters in order. This setting will set the default selected value in the create form. |
```
```suggestion
| default_x_axis_labels | "letters"  | "numbers" | Label style for the floor plan gird. Can use numbers or letters in order. This setting will set the default selected value in the create form. |
```

joewesch commented on nautobot_floor_plan/forms.py:
```diff
@@ -71,6 +85,8 @@ class FloorPlanTileForm(NautobotModelForm):
     rack = DynamicModelChoiceField(
         queryset=Rack.objects.all(), required=False, query_params={"nautobot_floor_plan_floor_plan": "$floor_plan"}
     )
+    x_origin = forms.CharField(validators=[])
+    y_origin = forms.CharField()
```
Is there a reason why `x_origin` has `validators=[]` and `y_origin` doesn't?

pszulczewski replied:
leftover ... removed

joewesch commented:
@glennmatthews would you mind giving this a once over?

pszulczewski commented on nautobot_floor_plan/forms.py:
```diff
@@ -87,3 +104,53 @@ class Meta:
+        """Common clean method for origin fields."""
+        value = self.cleaned_data.get(field_name)
+        if self.x_letters and field_name == "x_origin" or self.y_letters and field_name == "y_origin":
+            if self.letter_validator(field_name, value, axis) is not True:
+                return 0  # required to pass model clean() method
```
I was thinking to raise `ValidationError` in here, but what happens is that `ValidationError` exception is raised silently and it still goes to the model .clean() method, where it causes exception on the `>` operator `TypeError: '>' not supported between instances of 'str' and 'int'`, what results with an exception page in the UI.

pszulczewski commented on nautobot_floor_plan/forms.py:
```diff
@@ -87,3 +104,53 @@ class Meta:
+
+    def letter_validator(self, field, value, axis):
+        """Validate that origin uses combination of letters."""
+        if not re.search(r"[A-Z]+", str(value)):
```
`value.isupper()` and `value.isdigit()` !!! You're brilliant!

pszulczewski commented on nautobot_floor_plan/forms.py:
```diff
@@ -34,9 +38,18 @@ class Meta:
+        """Overwrite the constructor to set initial values for select widget."""
+        super().__init__(*args, **kwargs)
+        if not self.instance.created:
+            self.initial["x_axis_labels"] = get_app_settings_or_config("nautobot_floor_plan", "default_x_axis_labels")
+            self.initial["y_axis_labels"] = get_app_settings_or_config("nautobot_floor_plan", "default_y_axis_labels")
```
I think I could do that in the plugin `ready()` method and rise Django ConfigurationError? Would that be the right place?

glennmatthews replied:
Sounds worth a try to me.

pszulczewski replied:
Added in the latest commit.
